packages/contracts-bedrock/contracts/L1/L2OutputOracle.sol

```diff
@@ -204,59 +191,129 @@ contract L2OutputOracle is Initializable, Semver {
             // blockhash value, and delay submission until it is confident that the L1 block is
             // finalized.
             require(
-                blockhash(_l1BlockNumber) == _l1Blockhash,
+                blockhash(_l1BlockNumber) == _l1BlockHash,
                 "L2OutputOracle: blockhash does not match the hash at the expected height"
             );
         }
 
-        emit OutputProposed(_outputRoot, block.timestamp, _l2BlockNumber);
+        emit OutputProposed(_outputRoot, nextOutputIndex(), block.timestamp, _l2BlockNumber);
 
-        l2Outputs[_l2BlockNumber] = Types.OutputProposal(_outputRoot, block.timestamp);
-        latestBlockNumber = _l2BlockNumber;
+        l2Outputs.push(
+            Types.OutputProposal({
+                outputRoot: _outputRoot,
+                timestamp: uint128(block.timestamp),
+                l2BlockNumber: uint128(_l2BlockNumber)
+            })
+        );
     }
 
     /**
-     * @notice Returns the L2 output proposal associated with a target L2 block number. If the
-     *         L2 block number provided is between checkpoints, this function will rerutn the next
-     *         proposal for the next checkpoint.
-     *         Reverts if the output proposal is either not found, or predates
-     *         the startingBlockNumber.
+     * @notice Returns an output by index. Exists because Solidity's array access will return a
+     *         tuple instead of a struct.
      *
-     * @param _l2BlockNumber The L2 block number of the target block.
+     * @param _l2OutputIndex Index of the output to return.
+     *
+     * @return The output at the given index.
      */
-    function getL2Output(uint256 _l2BlockNumber)
+    function getL2Output(uint256 _l2OutputIndex)
         external
         view
         returns (Types.OutputProposal memory)
     {
+        return l2Outputs[_l2OutputIndex];
```

Comment:
Do we want our own error rather than `Panic(0x32)`?

Reply:
I think it's fine